SG14_test/inplace_function_test.cpp

```diff
@@ -429,6 +429,48 @@ static void RvalueRefParameter()
     g(std::make_unique<int>(42));
 }
 
+struct NoDefaultCtor
+{
+    int val;
+    explicit NoDefaultCtor(int v) : val{v} {}
+};
+
+int overloaded_function(stdext::inplace_function<int()>&& exec)
+{
+    return exec();
+}
+
+int overloaded_function(stdext::inplace_function<int(int)>&& exec)
+{
+    return exec(int{42});
+}
+
+int overloaded_function(stdext::inplace_function<int(int, NoDefaultCtor)>&& exec)
+{
+    return exec(int{42}, NoDefaultCtor{5});
+}
+
+int overloaded_function(stdext::inplace_function<void(int&)>&& exec)
+{
+    int out_parameter{88};
+    exec(out_parameter);
+    return out_parameter;
+}
+
+int fobau(NoDefaultCtor) { return 2; }
+
+
+void test_overloading()
+{
+    using brokenType = decltype(fobau(std::declval<NoDefaultCtor>()));
+    EXPECT_EQ(brokenType{2}, 2);
+
+    EXPECT_EQ(overloaded_function([]() -> int { return 3; }), 3);
+    EXPECT_EQ(overloaded_function([](int arg) -> int { return arg; }), 42);
+    EXPECT_EQ(overloaded_function([](int arg, NoDefaultCtor ndc) -> int { return arg + ndc.val; }), 42 + 5);
+    EXPECT_EQ(overloaded_function([](int& arg) -> void { arg += 3; }), 88 + 3);
```

Comment:
Personally, I would rather see these test cases as simple type-trait tests:

    auto a = []() -> int { return 3; };
    assert(( std::is_convertible<decltype(a), stdext::inplace_function<int()>>::value ));
    assert(( !std::is_convertible<decltype(a), stdext::inplace_function<int(int)>>::value ));
    assert(( !std::is_convertible<decltype(a), stdext::inplace_function<void(int&)>>::value  ));
    auto b = [](int&) -> void {};
    assert(( !std::is_convertible<decltype(b), stdext::inplace_function<int()>>::value ));
    assert(( !std::is_convertible<decltype(b), stdext::inplace_function<int(int)>>::value ));
    assert(( std::is_convertible<decltype(b), stdext::inplace_function<void(int&)>>::value ));

Reply:
Ok, are you sure that tests the exact same thing, as in can that cause an ambiguous overload error? Also why `assert` and not `static_assert`?

Reply:
Tests the same thing: Yes, in that if only exactly one of the `is_convertible`s is true, then of course the overload resolution won't be ambiguous. However, neither your original tests nor mine would handle (perhaps impossible-to-implement) things where both conversions are *possible* but have different overload-resolution priorities anyway, such as

    A foo(inplace_function<int()>);
    B foo(inplace_function<long()>);
    int bar();
    A a = foo(bar);  // test that IPF<int()> is a better match than IPF<long()>
    // std::function does not support this, and I think it might be impossible to support

    A foo(inplace_function<void(const int&)>);
    B foo(inplace_function<void(int&)>);
    void bar(const int&);
    A a = foo(bar);  // test that IPF<void(const int&)> is a better match than IPF<void(int&)>
    // std::function does not support this, and I think it might be impossible to support



`assert` instead of `static_assert`: brain fart. 😛

Reply:
https://cplusplus.github.io/LWG/issue2813 is related.